mapping_parenting_tech/analysis/prototyping/clustering_descriptions.py

```diff
@@ -57,45 +76,380 @@
+    },
+    "discard": {
+        0: "Russian",
+        1: "religion",
+        3: "driving test",
+        4: "web safety",
+        14: "outdoor / mapping",
+        15: "engineering/physics",
+        21: "medical education",
+        23: "medical 'organising'",
+        24: "medical",
+    },
+}
+
+# %%
+relevant_clusters = set()
+for a_id in set(extra_apps + apps_of_interest):
+    if a_id in df.appId.values:
+        relevant_clusters.update(df[df.appId == a_id].cluster.values)
+        if df[df.appId == a_id].cluster.values == -1:
+            print(f"{a_id} not in cluster")
+
+relevant_clusters
 
 # %%
 driver = google_chrome_driver_setup()
 
 # %%
-save_altair(fig, "cluster_descriptions", driver)
+# save the figure if so wished (Currently throws an error on MRH M1)
+# save_altair(fig, "cluster_descriptions", driver)
+
+# %% [markdown]
+# Following code repeats the steps above, but uses apps' summaries (rather than their descriptions) to do the clustering. This
+# provides a point of validation/confirmation
+
+# %%
+filename = "summary_embeddings-22-01-21.pickle"
+with open(INPUT_DATA / filename, "rb") as f:
+    summary_embeddings = pickle.load(f)
+
+# %%
+# Create a 2D embedding
+s_reducer = umap.UMAP(n_components=2, random_state=1)
+s_embedding = s_reducer.fit_transform(summary_embeddings)
+
+# %%
+# Check the shape of the reduced embedding array
+s_embedding.shape
+
+# %%
+# Create another low-dim embedding for clustering
+s_reducer_clustering = umap.UMAP(n_components=50, random_state=1)
+s_embedding_clustering = s_reducer_clustering.fit_transform(summary_embeddings)
+
+# %%
+# Clustering with hdbscan
+np.random.seed(3667)
+s_clusterer = hdbscan.HDBSCAN(
+    min_cluster_size=42,
+    min_samples=1,
+    cluster_selection_method="leaf",
+    prediction_data=True,
+)
+s_clusterer.fit(s_embedding_clustering)
+logging.info(
+    f"{len(set(s_clusterer.labels_))} clusters with {sum([bool(x) for x in s_clusterer.labels_ if int(x) == -1])} unassigned apps"
+)
+
+# %%
+# Prepare dataframe for visualisation
+s_df = details.copy()
+s_df["x"] = s_embedding[:, 0]
+s_df["y"] = s_embedding[:, 1]
+s_df["s_cluster"] = [int(x) for x in s_clusterer.labels_]
+s_df["s_cluster_confidence"] = s_clusterer.probabilities_
+
+# %%
+s_cluster_col = s_df.columns.get_loc("s_cluster")
+
+radius = 3
+neigh = KNeighborsClassifier(n_neighbors=radius)
+
+current_unassigned = len(s_df[s_df.s_cluster == -1].s_cluster.to_list())
+
+s_cluster_list = s_df.s_cluster.to_list()
+
+unassigned_indices = []
+unassigned_indices.extend([i for i, x in enumerate(s_cluster_list) if x == -1])
+
+X_list = [
+    x for i, x in enumerate(s_embedding_clustering) if i not in unassigned_indices
+]
+y_list = [i for i in s_cluster_list if i != -1]
+
+neigh.fit(X_list, y_list)
+
+for x in unassigned_indices:
+    s_df.iat[x, s_cluster_col] = neigh.predict([s_embedding_clustering[x]])
+
+print(f"{len(s_df[s_df.s_cluster == -1].s_cluster.to_list())} unassigned apps remain.")
+
```

Comment:
You can calculate the similarity between both clustering approaches like this:
```suggestion

# Check the correspendence between both clustering approaches
print(adjusted_mutual_info_score(s_df.s_cluster.to_list(), df.cluster.to_list()))
```

This returns `0.5589677514190944` for me, which indicates that there is quite considerable similarity, but also some significant differences (1 would mean that there is perfect similarity). FYI you could get more detailed understanding by using confusion matrices.